src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -196,18 +177,25 @@ async def mypy_typecheck_partition(
         CreateDigest([FileContent(file_list_path, "\n".join(python_files).encode())]),
     )
 
-    typechecked_venv_pex_request = Get(
+    # This creates a venv with all the 3rd-party requirements used by the code. We tell MyPy to
+    # use this venv by setting `--python-executable`. Note that this Python interpreter is
+    # different than what we run MyPy with.
+    #
+    # We could have directly asked the `PexFromTargetsRequest` to return a `VenvPex`, rather than
```

Comment:
If true, this sounds way too titchy. I assume I now have to be thinking about this sort of thing in a whole bunch of rules. Is this somehow local to this MyPy rule because of specialness here or is it a trap that ideally no-one would need to worry about if we did X?

Reply:
Agreed it's too titchy. It's a trap anyone could run into. If you did `await Get(VenvPex, PexFromTargetsRequest)`, it would go `PexFromTargetsRequest -> PexRequest -> VenvPex`. Whereas all the other rules currently are doing `PexFromTargetsRequest -> PexRequest -> Pex`. We want to make sure all the call sites are using the same type of resulting Pex so that they can leverage the same work.

I think the best way to fix this is newtype `PexFromTargets` to wrap a `Pex`? It will add to boilerplate, but force people into doing the right thing. I'll add this as a followup.